components/raftstore/src/store/peer.rs

```diff
@@ -1592,39 +1777,63 @@ where
                     "region_id" => self.region_id,
                     "peer_id" => self.peer.get_id(),
                 );
-                return None;
+                return false;
             }
             CheckApplyingSnapStatus::Success => {
                 fail_point!("raft_before_applying_snap_finished");
-                // 0 means snapshot is scheduled after being restarted.
-                if self.last_unpersisted_number != 0 {
-                    // Because we only handle raft ready when not applying snapshot, so following
-                    // line won't be called twice for the same snapshot.
+
+                if let Some(mut snap_ctx) = self.snap_ctx.take() {
+                    // This snapshot must be scheduled
+                    assert!(snap_ctx.scheduled);
+
+                    let raft_msgs = mem::take(&mut snap_ctx.msgs);
+                    self.send_raft_msg(ctx, raft_msgs);
+
+                    // Snapshot has been persisted.
+                    self.last_applying_idx = self.get_store().truncated_index();
                     self.raft_group.advance_apply_to(self.last_applying_idx);
                     self.cmd_epoch_checker.advance_apply(
                         self.last_applying_idx,
                         self.term(),
                         self.raft_group.store().region(),
                     );
                 }
+                // If `snap_ctx` is none, it means this snapshot does not
+                // come from the ready but comes from the unfinished snapshot task
+                // after restarting.
+
+                // Note that this function must be called after applied index is updated,
+                // i.e. call `RawNode::advance_apply_to`.
                 self.post_pending_read_index_on_replica(ctx);
                 // Resume `read_progress`
                 self.read_progress.resume();
                 // Update apply index to `last_applying_idx`
                 self.read_progress.update_applied(self.last_applying_idx);
-                if !self.pending_messages.is_empty() {
-                    let msgs = mem::take(&mut self.pending_messages);
-                    self.send(ctx, msgs);
-                }
             }
             CheckApplyingSnapStatus::Idle => {
                 // FIXME: It's possible that the snapshot applying task is canceled.
                 // Although it only happens when shutting down the store or destroying
-                // the peer, it's still dengerous if continue to handle ready for the
+                // the peer, it's still dangerous if continue to handle ready for the
                 // peer. So it's better to revoke `JOB_STATUS_CANCELLING` to ensure all
                 // started tasks can get finished correctly.
+                self.snap_ctx = None;
```

Comment:
Doesn't seem correct to me. How about returning false here if `snap_ctx` is `Some`?

Reply:
Done. 
Actually, it's the same as the original code, if the snapshot is canceled and `pending_remove` is false, we will continue to handle ready. 
If `snap_ctx` is `Some`, return false can fix part of this problem. Another case is the snapshot comes from restarting and the `snap_ctx` is `None`.